File: Codigos estudiantes por lenguaje/Java/Luciano_Chavarria/Algoritmo_de_Ordenamiento/TreeSort.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Main {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String entrada[] = br.readLine().split(" ");
        int arr[] = new int[entrada.length];
        for (int i = 0; i < entrada.length; i++) {
            arr[i] = Integer.parseInt(entrada[i]);
        }
        treeInsert(arr);
        inroderRec(root);
    }

    static class Nodo {

        int key;
        Nodo izquierda, derecha;

        public Nodo(int item) {
            this.key = item;
            izquierda = derecha = null;
        }

    }
    static Nodo root;

    static void init() {
        root = null;
    }

    static void insert(int key) {
        root = insertRec(root, key);
    }

    static Nodo insertRec(Nodo root, int key) {
        if (root == null) {
            root = new Nodo(key);
            return root;
        }

        if (key < root.key) {
            root.izquierda = insertRec(root.izquierda, key);
        } else if (key >= root.key) {
            root.derecha = insertRec(root.derecha, key);

        }
        return root;
    }

    static void inroderRec(Nodo root) {
        if (root != null) {
            inroderRec(root.izquierda);
            System.out.print(root.key + " ");
            inroderRec(root.derecha);
        }
    }

    static void treeInsert(int arr[]) {
        for (int i = 0; i < arr.length; i++) {
            insert(arr[i]);
        }
    }

}

File: Codigos estudiantes por lenguaje/PY/Bryann Valderrama/Algoritmos de Ordenamiento/treeSort.py
'''Tree Sort
   
   - Crea una Arbol de Busqueda Binario con los elementos del arreglo.
   - Luego realiza un recorrido entre:
        InOrder (Menor a Mayor)
        PosOrder(Mayor a Menor). 
   - Complejidad Tiempo: 
        Promedio: O(n log n)
        Peor Caso: O(n^2)
'''

from sys import stdin, stdout, maxsize
rl = stdin.readline
wr = stdout.write


class Hoja:

    def __init__(self, item):
        self.key = item
        self.right = None
        self.left = None


def init():
    global root
    root = None


def insert(key):
    global root
    root = insertRec(root, key)


def insertRec(raiz, key):
    global root

    if raiz == None:
        raiz = Hoja(key)
        return raiz

    if key < raiz.key:
        raiz.left = insertRec(raiz.left, key)
    elif key >= raiz.key:
        raiz.right = insertRec(raiz.right, key)

    return raiz


def inOrder(actual):
    if actual != None:
        inOrder(actual.left)
        wr(f'{actual.key} ')
        inOrder(actual.right)


def insertInTree(arr):
    for i in range(len(arr)):
        insert(arr[i])


arr = list(map(int, rl().split()))
root = Hoja(arr[0])
insertInTree(arr)
inOrder(root)
